blog_posts/2024-09-12-optimisations-text-more.markdown

```diff
@@ -0,0 +1,144 @@
+
+This came with a lot of other smaller improvements too; there's a focus rectangle around objects, many ActionScript events are more correctly implemented, and navigating objects using the arrow keys should work just fine too.
+
+This is another one of those areas that just felt annoying when it doesn't work, but you'd likely never think twice about it when it does!
+
+## Optimisations
+The priority for Ruffle has always been correctness above all else. We implement something, get it right, *make tests*, and then look at how to speed it up. Until recently, we've been firmly in the "get things working" stage. Well, if you look at our progress percentage above - I think you'll see that quite a lot of things work the way they're supposed to these days! That means it's time to visit all the pain points for performance, and oh boy there's been some huge improvements in this area.
+
+[@Lord-McSweeney](https://github.com/Lord-McSweeney) and [@adrian17](https://github.com/adrian17) have implemented a verifier for AVM2, which lets us safely make assumptions about code and optimise based on those. This has unlocked adding an optimiser pass, which greatly speeds up ActionScript 3 content and removes a lot of overhead.
+It's still far from Flash Player speeds, but they have a JIT and had decades to make everything as fast as it was. We're getting there, though!
+
+Some other notable optimisations:
+- [@adrian17](https://github.com/adrian17) has implemented dependent strings and extended our use of interned strings, reducing the amount of memory needed for string operations in ActionScript and speeding up lots of basic text processing.
+- [@adrian17](https://github.com/adrian17) has also improved caching in Stage3D content, giving us up to 4x the FPS on some content that uses it.
+- [@dinnerbone](https://github.com/dinnerbone) has optimised general rendering slightly, reducing the number of times we submit buffers to the GPU per frame.
+- [@kjarosh](https://github.com/kjarosh) has improved the speed of some rounding functions by about 20%. These functions are used all over the codebase all the time.
+- [@moulins](https://github.com/moulins) has been working on improvements to how Ruffle uses [gc-arena](https://github.com/kyren/gc-arena), which reduces a lot of general overhead to anything that can be garbage collected.
+- [@Lord-McSweeney](https://github.com/Lord-McSweeney) has been refactoring our implementation of an ActionScript 3 `class`, simplifying a lot of code and allowing for even more performance improvements.

```

Comment:
_an_ ActionScript 3 `class` vs _the_ ActionScript 3 `class`?

Reply:
This is where the English language gets a bit funny 😅
"the" `class` makes it sound like it's either the singular "actionscript 3" class, or the implementation of the `class` keyword.
"a" `class` makes it sound like it's anything that is a class, which I think is more accurate here?

Reply:
Oh! I thought you meant the `class` class implementation here, I wouldn't have guessed that you meant classes in general. Maybe `implementation of ActionScript 3 classes` or `implementation of ActionScript 3 class model`?

Reply:
I guess the first one might mean we're talking about some concrete AS classes (e.g. from the API)...

Reply:
How's it now?

Reply:
Perfect!